Add support for HLS streams in just_audio_web

dindils opened the pull request:
Hi, thanks for maintaining this great library.

I needed HLS support for a project I'm working on so this change adds support for HLS streams in just_audio_web based on the work done by liorshk as discussed in issue [#242](https://github.com/ryanheise/just_audio/issues/242).

This utilizes HLS.JS for playback, added note in README.md on the script tag that needs to be added to fetch that library if HLS playback is desired.

dindils committed:
Add HLS support for web

dindils committed:
Update changelog and increment minor version

dindils committed:
Fix audio source not updating when calling setAudioSource

dindils committed:
Adjust after static analysis

dindils committed:
Remove debug logs for HLS.js

ryanheise commented:
Thanks for this. Now first, you may (or may not) be aware that there is an existing PR #1246 by @aletorrado also using HLS.js, and so I would want to credit both of you in the CHANGELOG, although yours looks more up to date using extension types.

Can I get you to just get you to check #1246 and compare it with your own to see if it sparks any further changes to your own PR?

aletorrado commented:
Hi! Thanks @ryanheise for keeping my PR in mind.

I don't like the need to synchronously preload Hls.js, that would delay the initial load of the app. I'd prefer to lazy load Hls.js like in #1246. To avoid dependance on hard coded CDN's, I would just consider to self host the Hls.js bundle.

dindils committed:
Update CHANGELOG and recommend async script tag

dindils commented:
@ryanheise Was not aware of that PR, thanks for pointing that out. No real changes that I see a need for except using async in the script tag. Updated the CHANGELOG to reflect the contributions for this feature.

@aletorrado Really good point! I updated the README.md to recommend to people to use the async tag when loading hls.js which I can see is a good improvement. Although I like your solution to load hls.js automatically when needed, I think I'd prefer to keep it as it is since then I don't have to worry about any hard-coded dependencies since it's up to the user of the library to provide it. For my use case I'm only streaming HLS files so I'd prefer to load it immediately anyways.

aletorrado commented:
I don't think just adding the async attribute will be the best solution, because that would introduce race conditions. Think about an application that plays a media file automatically on startup (maybe an app for a radio station), but Hls.js may not have loaded at that point. Somehow, the dart code should wait for Hls.js to load.

Regarding the URL for the Hls.js bundle, maybe it should be configurable, while defaulting to the official CDN provided by Hls.js.

@ryanheise what do you think about self hosting the js bundle for Hls.js?

ryanheise commented:
@aletorrado , I think as long as the app developer can choose between self hosting and CDN, it should be fine. So I rather than bundle a copy of hls.js with the plugin, the README could just give instructions.

What I'll do to get you both in the change log is I'll create a new `feature/hls` branch incorporating @dindils 's extension types, then @aletorrado would you be able to make a PR against that branch to merge in your proposed changes?

Edit: done. Let's continue on this branch #1523

@dindils I have reverted your "new formatter" changes since we're only on sdk ^3.3.0 (and it cleans up the diffs).

aletorrado commented:
@ryanheise Ok, just to check: Do you want me to integrate my Hls.js loader while preserving the integration into Html5AudioPlayer from @dindils fork?

ryanheise commented:
@aletorrado sorry for the delay (had an emergency to deal with).

> Ok, just to check: Do you want me to integrate my Hls.js loader while preserving the integration into Html5AudioPlayer from @dindils fork?

Yes, exactly. We include @dindils 's extension type code combined with your loading code and then (and I guess the README example script tag if it needs to change). That basically means you would create a new branch based on `feature/hls` as the starting point, and I can merge that. I see @dindils has already put you in the CHANGELOG so basically once that's merged I'd be happy to test and release.
